File: problems/链表/21.合并两个有序链表.py
#!/usr/bin/env python3
"""
21. 合并两个有序链表 (Merge Two Sorted Lists)
题目: 将两个升序链表合并为一个新的升序链表并返回。新链表是通过拼接给定的两个链表的所有节点组成的。
"""

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def merge_two_lists(l1: ListNode, l2: ListNode) -> ListNode:
    dummy = ListNode(0)
    cur = dummy
    while l1 and l2:
        if l1.val <= l2.val:
            cur.next = l1
            l1 = l1.next
        else:
            cur.next = l2
            l2 = l2.next
        cur = cur.next
    cur.next = l1 or l2
    return dummy.next

def list_to_linkedlist(lst):
    dummy = ListNode(0)
    cur = dummy
    for val in lst:
        cur.next = ListNode(val)
        cur = cur.next
    return dummy.next

def linkedlist_to_list(node):
    result = []
    while node:
        result.append(node.val)
        node = node.next
    return result

if __name__ == "__main__":
    # 测试用例1: 1->2->4 和 1->3->4 合并为 1->1->2->3->4->4
    l1 = list_to_linkedlist([1, 2, 4])
    l2 = list_to_linkedlist([1, 3, 4])
    result = merge_two_lists(l1, l2)
    print("测试1 - 合并 [1,2,4] 和 [1,3,4]:")
    print(f"  输出: {linkedlist_to_list(result)}")  # [1, 1, 2, 3, 4, 4]

    # 测试用例2: 空链表和 0->1->2->3 合并为 0->1->2->3
    l1 = None
    l2 = list_to_linkedlist([0, 1, 2, 3])
    result = merge_two_lists(l1, l2)
    print("\n测试2 - 合并 [] 和 [0,1,2,3]:")
    print(f"  输出: {linkedlist_to_list(result)}")  # [0, 1, 2, 3]

    # 测试用例3: [] 和 [] 合并为 []
    l1 = None
    l2 = None
    result = merge_two_lists(l1, l2)
    print("\n测试3 - 合并 [] 和 []:")
    print(f"  输出: {linkedlist_to_list(result)}")  # []

File: problems/链表/21.合并两个有序链表.cpp
#include <iostream>
#include <vector>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
};

ListNode* listToLinkedList(vector<int> lst) {
    ListNode* dummy = new ListNode(0);
    ListNode* cur = dummy;
    for (int val : lst) {
        cur->next = new ListNode(val);
        cur = cur->next;
    }
    return dummy->next;
}

vector<int> linkedListToList(ListNode* node) {
    vector<int> result;
    while (node) {
        result.push_back(node->val);
        node = node->next;
    }
    return result;
}

ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
    ListNode* dummy = new ListNode(0);
    ListNode* cur = dummy;
    while (l1 && l2) {
        if (l1->val <= l2->val) {
            cur->next = l1;
            l1 = l1->next;
        } else {
            cur->next = l2;
            l2 = l2->next;
        }
        cur = cur->next;
    }
    cur->next = l1 ? l1 : l2;
    return dummy->next;
}

void printList(ListNode* node) {
    cout << "[";
    while (node) {
        cout << node->val;
        if (node->next) cout << ",";
        node = node->next;
    }
    cout << "]";
}

int main() {
    // 测试用例1: 1->2->4 和 1->3->4 合并为 1->1->2->3->4->4
    ListNode* l1 = listToLinkedList({1, 2, 4});
    ListNode* l2 = listToLinkedList({1, 3, 4});
    ListNode* result = mergeTwoLists(l1, l2);
    cout << "测试1 - 合并 [1,2,4] 和 [1,3,4]:" << endl;
    cout << "  输出: ";
    printList(result);
    cout << endl;

    // 测试用例2: 空链表和 0->1->2->3 合并为 0->1->2->3
    l1 = nullptr;
    l2 = listToLinkedList({0, 1, 2, 3});
    result = mergeTwoLists(l1, l2);
    cout << "\n测试2 - 合并 [] 和 [0,1,2,3]:" << endl;
    cout << "  输出: ";
    printList(result);
    cout << endl;

    // 测试用例3: [] 和 [] 合并为 []
    l1 = nullptr;
    l2 = nullptr;
    result = mergeTwoLists(l1, l2);
    cout << "\n测试3 - 合并 [] 和 []:" << endl;
    cout << "  输出: ";
    printList(result);
    cout << endl;

    return 0;
}
